### PR TITLE
add custom loss unittest

### PR DESCRIPTION
**Pull request by simanw304**

### CODE REVIEW

**Comment by simanw304** [comment]
cc @zhichao-li

File: pyzoo/test/zoo/pipeline/autograd/test_loss.py [hunk 1]
```diff
@@ -33,7 +33,7 @@ class TestLoss(ZooTestCase):
     # y_pred and y_true are all keras_tensor
     # zloss is a AbstractCriterion
     # The first dim of shape is batch
-    def compareLossWithKeras(self, kkloss_func, zloss, shape):
+    def compareLossWithKeras(self, kkloss_func, zloss, shape, sizeAverage):
```
Comment (zhichao-li) [inline pyzoo/test/zoo/pipeline/autograd/test_loss.py]: change to "sizeAverageKerasLoss=True"

File: pyzoo/test/zoo/pipeline/autograd/test_loss.py [hunk 2]
```diff
@@ -53,14 +53,50 @@ def compareLossWithKeras(self, kkloss_func, zloss, shape):
...
     def test_abs(self):
         def mean_absolute_error(y_true, y_pred):
             result = mean(abs(y_true - y_pred), axis=1)
             return result
         self.compareLossWithKeras(kloss.mean_absolute_error,
-                                  CustomLoss(mean_absolute_error, [3]), [2, 3])
+                                  CustomLoss(mean_absolute_error, [3]), [2, 3], sizeAverage=True)
+
+    def test_zloss(self):
+        def zloss(**kwargs):
```
Comment (zhichao-li) [inline pyzoo/test/zoo/pipeline/autograd/test_loss.py]: Give a name to the loss. `rank_hinge_loss` ?

File: pyzoo/test/zoo/pipeline/autograd/test_loss.py [hunk 3]
```diff
@@ -53,14 +53,50 @@ def compareLossWithKeras(self, kkloss_func, zloss, shape):
...
+                batch = kwargs['batch']
+
+            def _zloss(y_true, y_pred):
+                y_pred = y_pred * (y_true / y_true)
+                margin = 1.0
+                t1 = []
+                for y in range(0, batch, 2):
+                    x = y_pred.slice(0, y, 1)
+                    t1.append(x)
+                pos = merge(t1, mode="concat", concat_axis=0)
```
Comment (zhichao-li) [inline pyzoo/test/zoo/pipeline/autograd/test_loss.py]: `pos = merge([y_pred.slice(0, i, 1) for i in range(0, batch, 2)], mode="concat", concat_axis=0)`

**Comment by simanw304** [comment]
jenkins passed http://172.168.2.101:8080/view/ZOO-PR/job/ZOO-PR-Validation/761/

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/Narrow.scala [hunk 4]
```diff
@@ -63,16 +63,18 @@ class Narrow[T: ClassTag](
...
-    require(offset >= 0 && offset <= input(positiveDim) -1,
-      s"Invalid narrow offset for dim $dim: $offset, " +
-        s"offset should be within range [0, ${input(positiveDim) - 1}]")
+    require(positiveDim >= 0 && positiveDim <= input.length - 1,
+      s"Invalid select dim: $dim, dim should be within range [0, ${input.length - 1}]")
     val positiveLength = if (length < 0) length + input(positiveDim) - offset + 1 else length
-    require(positiveLength > 0 && positiveLength <= input(positiveDim) - offset,
-      s"Invalid narrow length for dim $dim with offset $offset: $length, " +
-        s"length should be within range (0, ${input(positiveDim) - offset}]")
+    //Cannot narrow the batch dimension, so we need to check dim value.
```
Comment (zhichao-li) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/Narrow.scala]: batch dimension is always -1 for now, so we skip the checking here.

**Comment by zhichao-li** [comment]
LGTM

**Comment by glorysdj** [comment]
http://172.168.2.101:8080/view/ZOO-PR/job/ZOO-PR-Validation/761/ not passed...
still has a ut failure

**Pull request merged**